mypy/specialtype.py

```diff
@@ -0,0 +1,912 @@
+from collections import OrderedDict
+
+from typing import List, Dict, Tuple, cast, Optional, Callable, TYPE_CHECKING
+
+from mypy.exprtotype import expr_to_unanalyzed_type, TypeTranslationError
+from mypy.nodes import (
+    TypeInfo, AssignmentStmt, FuncDef, ClassDef, Var, GDEF, Expression,
+    Block, NameExpr, TupleExpr, ListExpr, ExpressionStmt, PassStmt,
+    DictExpr, CallExpr, RefExpr, Context, SymbolTable, UNBOUND_TVAR,
+    MDEF, Decorator, TypeVarExpr, NewTypeExpr, StrExpr, BytesExpr,
+    ARG_POS, ARG_NAMED, ARG_NAMED_OPT, NamedTupleExpr, TypedDictExpr, Argument,
+    UnicodeExpr, EllipsisExpr, TempNode, EnumCallExpr,
+    COVARIANT, CONTRAVARIANT, INVARIANT, ARG_OPT, SymbolTableNode
+)
+from mypy.types import (
+    NoneTyp, CallableType, Instance, Type, TypeVarType, AnyType,
+    TypeVarDef, TupleType, UnboundType, TypedDictType, TypeType,
+)
+from mypy import join
+if TYPE_CHECKING:
+    import mypy.semanal
+
+
+class Special:
+    """
+    Groups special-cased types:
+    * NamedTuple
```

Comment:
I wanted this too, but I couldn't decide how to provide the services from SemanticAnalyzer without duplication. Like you say, inheritance (for code reuse) does not sound like a good idea.

I think there are two kinds of services: lookup operations and analysis. It might be possible to refactor the lookup to a separate class (SymbolTable/Scope).

I should give it more thought.